Client/Frontend/Browser/ToolbarVisibilityViewModel.swift

```diff
@@ -0,0 +1,318 @@
+  enum ToolbarState {
+    /// Any toolbars are expanded and fully visible. Scrolling to the top is available
+    case expanded
+    /// Any toolbars are collapsed
+    case collapsed
+    
+    /// The inverse of the current state
+    var inversed: ToolbarState {
+      switch self {
+      case .expanded: return .collapsed
+      case .collapsed: return .expanded
+      }
+    }
+    
+    /// Inverses the current toolbar state
+    mutating func inverse() {
+      self = inversed
+    }
+  }
+  
+  /// The current toolbar state based on previous actions sent
+  @Published var toolbarState: ToolbarState = .expanded {
+    didSet {
+      // Explicitly setting the state means progress should be reset
+      interactiveTransitionProgress = nil
+    }
+  }
+  
+  /// The SwiftUI animation to use when manually updating the toolbar state.
+  ///
+  /// This animation will be used by default when user actions end up inversing the state of the toolbar
+  var toolbarChangeAnimation: Animation {
+    .spring(response: 0.25, dampingFraction: 0.8, blendDuration: 0)
+  }
+  
+  /// A UIKit animation to use when manually updating the toolbar state
+  var toolbarChangePropertyAnimator: UIViewPropertyAnimator {
+    .init(duration: 0.25, dampingRatio: 1)
+  }
+  
+  /// The travel distance between the ``ToolbarState/expanded`` and ``ToolbarState/collapsed`` toolbar states
+  ///
+  /// This is how much the user has to pan before the current toolbar state inverses.
+  ///
+  /// This should be set to the height of the expanded toolbar, minus the height of a collapsed variant if
+  /// available. For bottom bar this should be the drag distance before a full state change (to match Safari
+  /// this will be the bottom safe area height)
+  var transitionDistance: CGFloat
+  
+  /// The progress between the current state and its inverse during an interactive transition such as
+  /// panning the scroll view
+  @Published private(set) var interactiveTransitionProgress: CGFloat?
+  
+  /// A minimum content height that can be set to stops the toolbar from collapsing.
+  ///
+  /// When this is set to `nil`, the current scroll view's content height and ``transitionDistance`` will
+  /// determine if the toolbar can collapse.
+  var minimumCollapsableContentHeight: CGFloat?
+  
+  /// A snapshot in time for a scrolling container.
+  ///
+  /// Contains some basic info required to handle toolbar visibility
+  struct ScrollViewSnapshot {
+    var contentOffset: CGPoint
+    var contentInset: UIEdgeInsets
+    var contentHeight: CGFloat
+    var frameHeight: CGFloat
+    var isDecelerating: Bool
+  }
+  
+  /// Vertical axis related information during a pan on a scroll view
+  struct PanState {
+    var yTranslation: CGFloat
+    var yVelocity: CGFloat
+  }
+  
+  /// Whether or not scrolling to the top of a scroll view is allowed when tapping the status bar
+  var isScrollToTopAllowed: Bool {
+    toolbarState == .expanded
```

Comment:
Nothing for now, once we have bottom url bar, it should always scroll to top by tapping on status bar, even in expanded state, this is what Safari does at least